crates/webidl/src/lib.rs

```diff
@@ -104,43 +109,70 @@ fn parse(webidl_source: &str) -> Result<backend::ast::Program> {
     Ok(program)
 }
 
-/// Compile the given WebIDL file into Rust source text containing
-/// `wasm-bindgen` bindings to the things described in the WebIDL.
-pub fn compile_file(webidl_path: &Path) -> Result<String> {
-    let ast = parse_file(webidl_path)?;
-    Ok(compile_ast(ast))
+fn retain<K: Copy + Ord, V>(
+    map: &mut BTreeMap<K, V>,
+    mut filter: impl FnMut(&K) -> bool,
+) {
+    let mut to_remove = Vec::new();
+    for k in map.keys() {
+        if !filter(k) {
+            to_remove.push(*k);
+        }
+    }
+    for k in to_remove {
+        map.remove(&k);
+    }
 }
 
 /// Compile the given WebIDL source text into Rust source text containing
 /// `wasm-bindgen` bindings to the things described in the WebIDL.
-pub fn compile(webidl_source: &str) -> Result<String> {
-    let ast = parse(webidl_source)?;
+pub fn compile(
+    webidl_source: &str,
+    allowed_types: Option<&[&str]>,
+) -> Result<String> {
+    let ast = parse(webidl_source, allowed_types)?;
     Ok(compile_ast(ast))
 }
 
-/// Run codegen on the AST to generate rust code.
-fn compile_ast(mut ast: backend::ast::Program) -> String {
-    // Iteratively prune all entries from the AST which reference undefined
-    // fields. Each pass may remove definitions of types and so we need to
-    // reexecute this pass to see if we need to keep removing types until we
-    // reach a steady state.
-    let builtin = BTreeSet::from_iter(
+fn builtin_idents() -> BTreeSet<Ident> {
+    BTreeSet::from_iter(
         vec![
             "str", "char", "bool", "JsValue", "u8", "i8", "u16", "i16", "u32", "i32", "u64", "i64",
             "usize", "isize", "f32", "f64", "Result", "String", "Vec", "Option",
             "ArrayBuffer", "Object", "Promise",
         ].into_iter()
             .map(|id| proc_macro2::Ident::new(id, proc_macro2::Span::call_site())),
-    );
+    )
+}
+
+/// Run codegen on the AST to generate rust code.
+fn compile_ast(mut ast: backend::ast::Program) -> String {
+    // Iteratively prune all entries from the AST which reference undefined
+    // fields. Each pass may remove definitions of types and so we need to
+    // reexecute this pass to see if we need to keep removing types until we
+    // reach a steady state.
```

Comment:
One might describe the pruning as a monotonic function, and iterating to a fixed point with this monotonic function as a "monotone framework"... ;)

Reply:
🎓